In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/sms-spam-collection-dataset/spam.csv


## Machine learning project of CREDIT CARD FRAUD DETECTION



# Importing libraries

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing,
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

## Data Collection and preprocessing

In [2]:
df = pd.read_csv('/kaggle/input/sms-spam-collection-dataset/spam.csv',encoding='latin-1')

In [3]:
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [4]:
column_drop=['Unnamed: 2','Unnamed: 3','Unnamed: 4']
df.drop(columns=column_drop, inplace=True)

In [5]:
df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [6]:
df.describe()

v1                      v2
count   5572                    5572
unique     2                    5169
top      ham  Sorry, I'll call later
freq    4825                      30

In [7]:
df.shape

(5572, 2)

In [8]:
## replace the null values with a null string
mail_data = df.where((pd.notnull(df)),'')

In [9]:
mail_data.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [10]:
mail_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   v1      5572 non-null   object
 1   v2      5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [11]:
mail_data.shape

(5572, 2)

# label encoding

In [12]:
## label spam mail as 0 ; and ham mail as 1
mail_data.loc[mail_data['v1']=='spam','v1',]=0
mail_data.loc[mail_data['v1']=='ham','v1',]=1

spam=0
ham=1

In [13]:
#seperating the data as texts and labels
X = mail_data['v2']
Y = mail_data['v1']

In [14]:
print(X)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                Will Ì_ b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: v2, Length: 5572, dtype: object


In [19]:
print(Y)

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: v1, Length: 5572, dtype: object


# spliting of data into train test

In [20]:
X_train, X_test, Y_train, Y_test = train_test_split (X, Y, test_size=0.2, random_state=3)

In [21]:
print(X.shape)
print(X_train.shape)
print(X_test.shape)

(5572,)
(4457,)
(1115,)


# Feature Extraction

In [22]:
feature_extraction = TfidfVectorizer()

X_train_features = feature_extraction.fit_transform(X_train)
X_test_features = feature_extraction.transform(X_test)

Y_train = Y_train.astype('int')
Y_test = Y_test.astype('int')

In [23]:
print(X_test_features)

  (0, 7720)	0.18177353326505832
  (0, 7529)	0.3065337254472181
  (0, 6200)	0.4854655760255235
  (0, 4429)	0.46634315006427723
  (0, 3646)	0.2366539368537363
  (0, 1586)	0.6031895330837128
  (1, 7720)	0.0680714933413295
  (1, 7462)	0.21219328500890222
  (1, 7426)	0.11457201769448462
  (1, 6805)	0.1376129736715364
  (1, 6785)	0.2515532326811641
  (1, 6502)	0.13930309431681578
  (1, 5682)	0.26383622694260933
  (1, 5525)	0.2515532326811641
  (1, 5428)	0.26383622694260933
  (1, 4994)	0.11009098573254651
  (1, 4174)	0.23055531529961845
  (1, 3476)	0.26383622694260933
  (1, 3404)	0.3432134171711547
  (1, 3360)	0.10236228487449013
  (1, 2991)	0.12752083197274744
  (1, 2944)	0.0966319945046976
  (1, 1615)	0.10952824030062666
  (1, 1592)	0.10030649242498009
  (1, 1279)	0.14720142914662773
  :	:
  (1113, 5988)	0.3792643940590298
  (1113, 4092)	0.3036313322736912
  (1113, 3767)	0.1827989161083793
  (1113, 3604)	0.21530402265168153
  (1113, 3360)	0.1914340903760714
  (1113, 3339)	0.3792643940590298

In [24]:
print(X_train_features)

  (0, 741)	0.28307455118083463
  (0, 3360)	0.1327523238442287
  (0, 4108)	0.21196015023008544
  (0, 4908)	0.12973225055917514
  (0, 3042)	0.26300555749396887
  (0, 946)	0.1151770452043338
  (0, 7464)	0.19261204588580316
  (0, 4431)	0.3421657916670175
  (0, 6805)	0.17846848640014898
  (0, 6873)	0.15216101010779184
  (0, 3497)	0.28307455118083463
  (0, 2178)	0.33952544349598
  (0, 3235)	0.3869898904365042
  (0, 3365)	0.22753426247664568
  (0, 1032)	0.1361275560580212
  (0, 7720)	0.1765620792792692
  (0, 3491)	0.19174855251416806
  (0, 4655)	0.2558426236041184
  (1, 4190)	0.3725861907992424
  (1, 7099)	0.42172200036894236
  (1, 6620)	0.46707907862382136
  (1, 3646)	0.2020333473623602
  (1, 6645)	0.553594666958471
  (1, 7704)	0.3433404875792393
  (2, 954)	0.4257390912308466
  :	:
  (4455, 7402)	0.15130978849620824
  (4455, 6316)	0.16857953235415776
  (4455, 6850)	0.14840315498751144
  (4455, 1592)	0.11611242093594701
  (4455, 6991)	0.1743411711262433
  (4455, 4647)	0.1711510506728716
  (44

# Training of Model

In [25]:
## logistic regression
model= LogisticRegression()

In [26]:
# training the logistic regression model  with training data
model.fit(X_train_features ,Y_train)

LogisticRegression()

## Evaluting the training data

In [28]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, f1_score

In [29]:
gnb = GaussianNB()
mnb = MultinomialNB()
bnb = BernoulliNB()

In [30]:
# prediction on training data 
prediction_on_training_data = model.predict(X_train_features)
accuracy_on_training_data = accuracy_score(Y_train, prediction_on_training_data)

In [31]:
print('Accuracy on training data : ',accuracy_on_training_data)

Accuracy on training data :  0.9739735247924612


In [32]:
# prediction on testing data 
prediction_on_test_data = model.predict(X_test_features)
accuracy_on_test_data = accuracy_score(Y_test, prediction_on_test_data)

In [33]:
print('Accuracy on testing data : ',accuracy_on_test_data)

Accuracy on testing data :  0.9757847533632287


In [36]:
# prediction on training data  calculating precision
prediction_on_training_data = model.predict(X_train_features)
precision_on_training_data = precision_score(Y_train, prediction_on_training_data)

In [37]:
print('precision on training data : ',precision_on_training_data)

precision on training data :  0.9715723270440252


In [38]:
# prediction on training data  calculating precision
prediction_on_test_data = model.predict(X_test_features)
precision_on_test_data = precision_score(Y_test, prediction_on_test_data)

In [39]:
print('precision on testing data : ',precision_on_test_data)

precision on testing data :  0.9726443768996961


## Conclusion
accuracy of  training data is 97%

precision of training data is 97%

acuuracy of testing data is 97%

precision of testing data is 97%